src/de/raw.rs

```diff
@@ -393,41 +225,53 @@ impl<'de, 'a> serde::de::Deserializer<'de> for &'a mut Deserializer<'de> {
     where
         V: serde::de::Visitor<'de>,
     {
-        match self.current_type {
+        match self.element.element_type() {
             ElementType::Null => visitor.visit_none(),
             _ => visitor.visit_some(self),
         }
     }
 
     fn deserialize_enum<V>(
         self,
-        _name: &str,
+        _name: &'static str,
         _variants: &'static [&'static str],
         visitor: V,
-    ) -> Result<V::Value>
+    ) -> std::result::Result<V::Value, Self::Error>
     where
         V: serde::de::Visitor<'de>,
     {
-        match self.current_type {
-            ElementType::String => visitor.visit_enum(self.deserialize_str()?.into_deserializer()),
+        match self.element.element_type() {
+            ElementType::String => visitor.visit_enum(self.get_string()?.into_deserializer()),
             ElementType::EmbeddedDocument => {
-                self.access_document(|access| visitor.visit_enum(access))
+                let doc = match self.value()? {
+                    RawBsonRef::Document(doc) => doc,
+                    _ => {
+                        return Err(Error::deserialization(
+                            "internal error: unexpected non-document",
```

Comment:
There are a few places where I had to add this kind of error; these should never happen in practice because the `value()` match is always immediately preceded by an `element_type()` check.